added ansible-deployment-example.yml file

d-rupp opened the pull request:
This shows how an awx-operator deployment can be done via a ansible playbook.

##### SUMMARY
This shows how an awx-operator deployment can be done via a ansible playbook.

For #986

##### ADDITIONAL INFORMATION
This can be improved in a lot of ways with more variables and all that, but it is a simple example that worked for us.

d-rupp committed:
added ansible-deployment-example.yml file
This shows how an awx-operator deployment can be done via a ansible playbook.

mhalano commented:
Very cool. Thank you. I made one myself that have, I think, some improvements. Here it is: https://gist.github.com/mhalano/e1b3f05e47e428c5739ad87b46a36ddc

mhalano commented on playbooks/ansible-deployment-example.yml:
```diff
@@ -0,0 +1,104 @@
+    dest: "~/awx-operator/kustomization.yaml"
+    mode: 0640
+
+- name: Deploy controller to the K8s cluster specified in ~/.kube/config.
+  shell: kustomize build . | kubectl apply -f -
+  environment:
+    PATH: "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin:/snap/bin"
+  args:
+    chdir: "~/awx-operator"
+  changed_when: true
```
You could use [Kustomize lookup plugin](https://docs.ansible.com/ansible/latest/collections/kubernetes/core/kustomize_lookup.html) here

d-rupp replied:
Thanks, implemented. Did not know that one.

mhalano commented on playbooks/ansible-deployment-example.yml:
```diff
@@ -0,0 +1,104 @@
+      apiVersion: kustomize.config.k8s.io/v1beta1
+      kind: Kustomization
+      resources:
+        # Find the latest tag here: https://github.com/ansible/awx-operator/releases
+        - github.com/ansible/awx-operator/config/default?ref={{ awx_operator_version }}
```
These lines download the AWX Operator directly, so it's redundant to download with Git.

d-rupp replied:
Removed the git checkout part, thanks!

d-rupp committed:
Converted to real playbook
implemented suggestions
- look up latest version and use that
- Git checkout removed because it is unnecessary
- deploy kustomization with kustomize plugin lookup
- made some more variables

rooftopcellist commented:
Hi @d-rupp, I like your idea of contributing back a playbook to install the operator and deploy an example AWX instance.  

FYI, we already have a playbook [here](https://github.com/ansible/awx-operator/blob/devel/ansible/instantiate-awx-deployment.yml) that just deploys the AWX instance.  

In my mind, there are 2 general use cases:
* Build and Deploy Operator to a namespace & Deploy AWX instance as well
* Only deploy AWX instance (assumes the operator is already available)

Would you be open to having these be separate playbook, but have the one that does both just do an `include_tasks: instantiate-awx-deployment.yml` task?

rooftopcellist commented:
Could you also move this to `./ansible` instead of `./playbooks`?  The `./playbooks` directory is somewhat managed by the operator-sdk, and may be modified when doing operator-sdk upgrades.  To avoid any issues, we have been putting playbooks like this in the `./ansible` directory.

d-rupp committed:
Suggestions implemented
- split into separate playbooks
- uses instantiate-awx-deployment.yml
- added some more defaults and examples

d-rupp commented:
@rooftopcellist suggestions implemented, please check if this looks like what you were thinking about.

d-rupp commented on ansible/ansible-deployment-example.yml:
```diff
@@ -0,0 +1,19 @@
+      include_tasks: deploy-awx-operator-kustomization.yml
+
+    - name: Deploy AWX Instance
+      include_tasks: instantiate-awx-deployment.yml
```
```suggestion
      include_tasks: instantiate-awx-deployment.yml
    
```

d-rupp committed:
Update ansible/ansible-deployment-example.yml

fix eof

d-rupp committed:
fix file ending again

d-rupp committed:
linter fixes

rooftopcellist commented on ansible/deploy-awx-operator-kustomization.yml:
```diff
@@ -0,0 +1,35 @@
+
+    - name: Create kubernetes resources for lookup output
+      kubernetes.core.k8s:
+        definition: "{{ lookup('kubernetes.core.kustomize', dir='.') }}"
```
When testing this out locally, I get this error:

```
TASK [Create kubernetes resources for lookup output] ****************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "lookup plugin (kubernetes.core.kustomize) not found"}
```

Looks like that's because I have kubernetes.core 1.2.1.  (https://docs.ansible.com/ansible/latest/collections/kubernetes/core/kustomize_lookup.html)


I will see if it is possible to bump our requirements.yml file for `kubernetes.core>=2.3.2`. I think it would be weird to expect users to have a newer kubernetes.core version that what we use in the operator roles.

rooftopcellist commented:
@d-rupp 

```
# Deploy AWX
ansible-playbook ansible/instantiate-awx-deployment.yml -e namespace=$NAMESPACE -e image=quay.io/ansible/awx -e image_version=latest -e service_type=clusterip -e ingress_type=route

# Deploy operator using kustomization
ansible-playbook ansible/deploy-awx-operator-kustomization.yml -e awx_namespace=ca-awx

# Deploy both using wrapper playbook
ansible-playbook ansible/ansible-deployment-example.yml 
```

This last one didn't work.  I think it's because the included files have hosts specified..

```
TASK [Deploy AWX Operator] ******************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"reason": "conflicting action statements: hosts, tasks\n\nThe error appears to be in '/home/chadams/awx-operator/ansible/deploy-awx-operator-kustomization.yml': line 2, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n---\n- hosts: localhost\n  ^ here\n"}
```

rooftopcellist commented:
@d-rupp did you get a chance to look at that last comment?

d-rupp committed:
requested changes

d-rupp commented:
> @d-rupp did you get a chance to look at that last comment?

Sorry, this got lost in the shuffle. You are correct, i removed that. Downside is you can not call them separately anymore of course.

rooftopcellist commented on ansible/ansible-deployment-example.yml:
```diff
@@ -0,0 +1,19 @@
+
+  tasks:
+    - name: Deploy AWX Operator
+      include_tasks: deploy-awx-operator-kustomization.yml
```
We could add an extra variable called `operator_only=true`, and have a `when:` here.  

Same for the AWX CR below.
